lib/msal-common/src/app/module/AuthorizationCodeModule.ts

```diff
@@ -107,77 +115,151 @@ export class AuthorizationCodeModule extends AuthModule {
-        }
+    async acquireToken(codeResponse: CodeResponse): Promise<TokenResponse> {
+        try {
+            const tokenRequest: TokenExchangeParameters = this.getCachedRequest();
 
-        const acquireTokenAuthority = (request && request.authority) ? AuthorityFactory.createInstance(request.authority, this.networkClient) : this.defaultAuthorityInstance;
-        if (!acquireTokenAuthority.discoveryComplete()) {
-            try {
-                await acquireTokenAuthority.resolveEndpointsAsync();
-            } catch (e) {
-                this.cacheManager.resetTempCacheItems(codeResponse.userRequestState);
-                throw ClientAuthError.createEndpointDiscoveryIncompleteError(e);
+            if(!codeResponse || !codeResponse.code) {
+                throw ClientAuthError.createTokenRequestCannotBeMadeError();
             }
-        }
 
-        const { tokenEndpoint } = acquireTokenAuthority;
-        let tokenReqParams;
-        try {
-            const tokenRequest: TokenExchangeParameters = request || this.getCachedRequest();
-            tokenReqParams = new ServerTokenRequestParameters(
+            const authorityKey: string = this.cacheManager.generateAuthorityKey(codeResponse.userRequestState);
+            const cachedAuthority: string = this.cacheStorage.getItem(authorityKey);
+            tokenRequest.authority = cachedAuthority;
+
+            const acquireTokenAuthority = (tokenRequest && tokenRequest.authority) ? AuthorityFactory.createInstance(tokenRequest.authority, this.networkClient) : this.defaultAuthorityInstance;
+            if (!acquireTokenAuthority.discoveryComplete()) {
+                try {
+                    await acquireTokenAuthority.resolveEndpointsAsync();
+                } catch (e) {
+                    throw ClientAuthError.createEndpointDiscoveryIncompleteError(e);
+                }
+            }
+
+            const { tokenEndpoint } = acquireTokenAuthority;
+            const tokenReqParams = new ServerTokenRequestParameters(
                 this.clientConfig.auth.clientId,
                 tokenRequest,
                 codeResponse,
+                this.getAccount(),
                 this.getRedirectUri(),
                 this.cryptoObj
             );
 
-            const acquiredTokenResponse = await this.networkClient.sendPostRequestAsync<ServerAuthorizationTokenResponse>(
-                tokenEndpoint,
-                {
-                    body: tokenReqParams.createRequestBody(),
-                    headers: tokenReqParams.createRequestHeaders()
+            return this.getTokenResponse(tokenEndpoint, tokenReqParams, tokenRequest, codeResponse);
+        } catch (e) {
+            this.cacheManager.resetTempCacheItems(codeResponse && codeResponse.userRequestState);
+            this.account = null;
+            throw e;
+        }
+    }
+
+    async renewToken(request: TokenRenewParameters): Promise<TokenResponse> {
+        try {
+            if (!request) {
```

Comment:
I would think you would want to check if this is an empty object (which would not get caught be this) as well, yeah?

Reply:
I think it is ok if it is an empty object, since we define all the object fields as optional. We have other checks for scopes and such, so not necessary to do here.

Reply:
I've rearranged the function so at least scopes will be checked before any execution is performed.